src/QuadForm/CloseVectors.jl

```diff
@@ -0,0 +1,178 @@
+export close_vectors
+
+@doc Markdown.doc"""
+    close_vectors(L:ZLat, v:Vector, upperbound;
+                  equal::Bool=false, check::Bool = false, filter)
+                                            -> Vector{Tuple{Vector{fmpz}}, fmpq}
+
+Return all tuples `(x, d)` where `x` is an element of `L` such that `d = b(v -
+x, v - x) <= c`.
+
+If `equal` is `true`, then only elements with `b(v - x, v - x) = c` are
+returned.
+
+If `filter` is not `nothing`, then only those `x` with `filter(x)` evaluating
+to `true` are returned.
+
+By default, it will be checked whether `L` is positive definite. This can be
+disabled setting `check = false`.
+
+Both input and output are with respect to the basis matrix of `L`.
+
+```jldoctest
+julia> L = Zlattice(matrix(QQ, 2, 2, [1, 0, 0, 2]));
+
+julia> close_vectors(L, [1, 1], 1)
+3-element Vector{Tuple{Vector{fmpz}, fmpq}}:
+ ([2, 1], 1)
+ ([0, 1], 1)
+ ([1, 1], 0)
+
+julia> close_vectors(L, [1, 1], 1, equal = true, filter = x -> all(>=(1), x))
+1-element Vector{Tuple{Vector{fmpz}, fmpq}}:
+ ([2, 1], 1)
+```
+"""
+function close_vectors(L::ZLat, v::Vector, upperbound; kw...)
+  return _close_vectors(L, QQ.(v), QQ(upperbound); kw...)
+end
+
+function _close_vectors(L::ZLat, v::Vector{fmpq}, upperbound::fmpq;
+                                equal::Bool=false, sorting::Bool=false,
+                                check=true, filter = nothing) where T <: RingElem
+  epsilon = QQ(1//10)   # some number > 0, not sure how it influences performance
+  d = length(v)
+  G1 = gram_matrix(rational_span(L))
+  if check
+    @req is_definite(L) == true && G1[1, 1] > 0 "Zlattice must be positive definite"
+    @req rank(L) == d "Zlattice must have the same rank as the length of the vector in the second argument."
+  end
+  # Construct
+  # G = [ G1 | 0 ]
+  #     [  0 | e ]
+  # where e = upperbound//3 + epsilon
+  G = zero_matrix(QQ, d + 1, d + 1)
+  _copy_matrix_into_matrix(G, 1, 1, G1)
+  e = upperbound//3 + epsilon
+  G[d + 1, d + 1] = e
+  # Construct
+  # B = [ 1  | 0 ]
+  #     [ -v | 1 ]
+  # since we will be working in the rational_span(L) + Qw with w^2 = e
+  B = identity_matrix(QQ, d + 1)
+  GC.@preserve B begin
+    for i in 1:d
+      m = ccall((:fmpq_mat_entry, libflint),
+                Ptr{fmpq}, (Ref{fmpq_mat}, Int, Int), B, d, i - 1)
+      ccall((:fmpq_set, libflint), Cvoid, (Ptr{fmpq}, Ref{fmpq}), m, v[i])
+      ccall((:fmpq_neg, libflint), Cvoid, (Ptr{fmpq}, Ptr{fmpq}), m, m)
+    end
+  end
+
+  N = Zlattice(B, gram = G, check = false)
+
+  delta = QQ(4//3)*upperbound + epsilon
+  sv = short_vectors(N, delta)
+  cv = Vector{Tuple{Vector{fmpz}, fmpq}}()
+  V = rational_span(L)
+  for a in sv
+    _a, _l = a
+    al = _a[end]
+    if iszero(al)
+      continue
+    end
+    @assert al == 1 || al == -1
+
+    x = fmpz[_a[i] for i in 1:d]
```

Comment:
What is your reason for this conversion? We are not doing it for `short_vectors` so why here?

Reply:
It was there before and I guess you had a reason. I will adjust it.